pkg/importer/imageio-datasource.go

```diff
@@ -108,24 +110,54 @@ func (is *ImageioDataSource) Info() (ProcessingPhase, error) {
 
 // Transfer is called to transfer the data from the source to a scratch location.
 func (is *ImageioDataSource) Transfer(path string) (ProcessingPhase, error) {
+	defer func() {
+		if err := cleanupTransfer(is.connection, is.imageTransfer); err == nil {
+			is.imageTransfer = nil // If successful, this avoids an unnecessary cleanup in Close
+		}
+	}()
 	// we know that there won't be archives
 	size, _ := util.GetAvailableSpace(path)
 	if size <= int64(0) {
 		//Path provided is invalid.
 		return ProcessingPhaseError, ErrInvalidPath
 	}
+	is.readers.StartProgressUpdate()
 	file := filepath.Join(path, tempFile)
 	err := util.StreamDataToFile(is.readers.TopReader(), file)
 	if err != nil {
 		return ProcessingPhaseError, err
 	}
 	// If we successfully wrote to the file, then the parse will succeed.
 	is.url, _ = url.Parse(file)
+
+	// Make sure the snapshot's backing file actually matches the expected parent snapshot.
+	// Otherwise, it is not safe to rebase the snapshot onto the previously-downloaded image.
+	// Need to check this after the transfer because it is not provided by the oVirt API.
+	if is.IsDeltaCopy() {
+		imageInfo, err := qemuOperations.Info(is.url)
+		if err != nil {
+			return ProcessingPhaseError, err
+		}
+
+		backingFile := filepath.Base(imageInfo.BackingFile)
+		if backingFile != is.previousSnapshot {
+			return ProcessingPhaseError, errors.Errorf("snapshot backing file '%s' does not match expected checkpoint '%s', unable to safely rebase snapshot", backingFile, is.previousSnapshot)
+		}
+
+		klog.Info("Successfully copied snapshot data, moving to merge phase.")
+		return ProcessingPhaseMergeDelta, nil
```

Comment:
Does it make sense to push this MergeDeltaPhase into the data processor. Are there other sources that would benefit from having the ability to MergeDelta? If not why not just merge the delta in this datasource?

Reply:
I kind of liked it this way better at first because it lined up the qemu-img rebase and commit operations with the existing data processor info/resize/convert code, but maybe the implementation is too specific to ImageIO? I'm not sure if there will be other data sources that would benefit from exactly this sequence.

Here is a version with the MergeDelta work rolled up into the ImageIO data source Transfer:
https://github.com/mrnold/containerized-data-importer/compare/imageio-snapshots...mrnold:imageio-snapshots-mergedelta?expand=1

I can merge that in to this pull request if it's preferable.

Reply:
Well my first reaction was hmm this seems very specific to imageio, so I don't think it makes sense to expose it in the data processor. But since it is not part of the datasource interface and thus not all the datasources have to implement it, it doesn't do much harm in the data processor. I just wanted to think about if there would be anything else that would have a similar need that would make better to have it in the data processor. So far I can't think of anything. The vddk importer has all of that inside the nbdkit filter.

Reply:
I am not that invested one way or the other, we can keep it as is for now, and see if there are other sources that will want it the same way (glance maybe?) if not we can always switch it.

Reply:
Glance might actually be a good guess, it appears to allow the use of qcow2 for at least some operations. It may very well be possible to do the same process for warm migration from OpenStack.